common-lib/src/cmr/common/xml.clj

```diff
@@ -163,21 +172,67 @@
+    (.setFeature spf "http://xml.org/sax/features/external-general-entities" false)
+    (.setFeature spf "http://xml.org/sax/features/external-parameter-entities" false)
+    ;; Enable secure processing feature
+    (.setFeature spf XMLConstants/FEATURE_SECURE_PROCESSING true)
+    spf))
+
 (defn validate-xml
   "Validates the XML against the schema in the given resource. schema-resource should be a classpath
   resource as returned by clojure.java.io/resource.
-  Returns a list of errors in the XML schema."
+  Returns a list of errors in the XML schema.
+  DTD validation is turned off and an error message is given."
   [^java.net.URL schema-resource xml]
-  (let [^SchemaFactory factory (SchemaFactory/newInstance XMLConstants/W3C_XML_SCHEMA_NS_URI)
-        schema (.newSchema factory schema-resource)
-        validator (.newValidator schema)
+  (let [spf (create-and-set-sax-parse-features)
+        ;; Configure the SchemaFactory and apply Schema to the SAXParserFactory
+        schema-factory (SchemaFactory/newInstance XMLConstants/W3C_XML_SCHEMA_NS_URI)
+        _ (.setProperty schema-factory XMLConstants/ACCESS_EXTERNAL_DTD "")
+        schema (try
+                 ;; Loads the schema as a file url
+                 (.newSchema schema-factory schema-resource)
+                 (catch Exception e
+                   (warn "Failed to load schema as URL, attempting as string - this is normal for a unit test:" (.getMessage e))
+                   ;; Loads the schema as a string needed for unit test
+                   (.newSchema schema-factory (StreamSource. (StringReader. schema-resource)))))
+        sax-parser (.newSAXParser spf)
+        xml-reader (.getXMLReader sax-parser)
+        ;; Set a "blanking" EntityResolver on the XMLReader
+        ;; This ensures that even if a DOCTYPE is present, the resolver prevents fetching any external DTD files.
+        _ (.setEntityResolver xml-reader
+                              (reify EntityResolver
+                                (resolveEntity [_ _public-id _system-id]
+                                  (InputSource. (StringReader. "")))))
+        ;; Create a SAXSource with the configured XMLReader.
+        input-source (InputSource. (StringReader. xml))
+        sax-source   (SAXSource. xml-reader input-source)
+        validator    (.newValidator schema)
         errors-atom (atom [])]
-    (.setErrorHandler validator (create-error-handler errors-atom))
     (try
-      (.validate validator (StreamSource. (StringReader. xml)))
+      (.setErrorHandler validator (create-error-handler errors-atom))
+      (.validate validator sax-source)
       (catch SAXParseException e
         ;; An exception can be thrown if it is completely invalid XML.
-        (reset! errors-atom [(sax-parse-exception->str e)])))
+        (reset! errors-atom [(sax-parse-exception->str e)]))
+      (catch SAXException e
+        (reset! errors-atom [(sax-parse-exception->str e)]))
+      (catch Exception e
```

Comment:
is the broad Exception catch(few in this PR) deliberate? Or could we be more precise here?